### DISCUSSION TITLE
Emulate FB10 for 2nd heating circuit

### DISCUSSION THREAD

**Discussion by vtamac** [emsesp/EMS-ESP32#1602] [Q&A]
Hi Everyone,

My current setup is as follows:

- Junkers ZSB 22-3 C 23
- FW 200 controller
- Gateway S3 from bbqkees

My building is a three-floor house (basement, living floor 1, living floor 2) with the boiler in the basement and the FW200 in the living room of floor 1. Each floor has its own underfloor heating circuit, but unfortunately, only a single FW200 is available to measure temperature and control the boiler and its single circulation pump. I am adding three more pumps for the three floors, and I have a bunch (10+) of temperature sensors across all floors integrated into Home Assistant.

Since the FW200 only supports two types of remote sensors, FB10 and FB100, both of which are discontinued, I believe my only option would be to emulate two (or three, and move the FW200 to the boiler) FB10(0) sensors. This way, the FW200 can remain the master thermostat, able to read desired/current temperatures from the remote ones and control the boiler accordingly.

Is it possible to solve this with the Gateway S3, even if it requires additional hardware? I noticed the menu "Custom Entities," but the documentation is pretty much empty about that.

I am open to any other additional ideas. Of course, the reason for the suggested approach above is to keep the FW200 with all its intelligence (some sort of PID controller inside, I guess). Let me know if a more convenient way exists.

Thank you,
Tamás

**Comment by MichaelDvP** [comment]
We can try, i've seen a log with a remote FB10 (product-id 109, ver 16.5) on hc1, sending direct to master with type-id 0x0123. For other hc is it unclear if they use the same telegram from different devices (like RC20 remotes) or different telegrams from different devices (like RC100H) We have to test. I'll prepare something in my testbuild.

**Reply by MichaelDvP** [reply]
Ok, first test with fixed type-id: https://github.com/MichaelDvP/EMS-ESP32/releases/tag/test
Please check. If it works for hc1 and not for hc2 i'll change hc2 type-id. If not working for hc1 i'll need a full log.

As for other remote emulation: To activate send temperaturevalue to thermostat/hc1/remotetemp (or in terminal `call thermostat hc1/remotetemp xx` with xx temperature or `call thermostat remotetemp xx 1`). You should get a new thermostat FB10 at device-id 18 (19 for hc2) with a single temperaturevalue (after a minute).  (Set thermostat control to FB10 to use roomtemp). To deactivate the remote send remotetemp value: `-1`

**Reply by vtamac** [reply]
Hi @MichaelDvP 

So I've uploaded the new FW to the S3 and executed the following (when the boiler was not heating as roomtemp on fw200 was higher than set)
<img width="1449" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/ae03ce53-f0d9-44ff-a40a-99603f1fe8dc">

then in telnet
`
ems-esp:# call thermostat hc1/remotetemp 11
000+00:02:34.385 I 5: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 11 and id 1 on device 0x10
ems-esp:# call thermostat hc2/remotetemp 11
Bad syntax (error code 5)
000+00:02:40.631 I 6: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 11 and id 2 on device 0x00
000+00:02:40.631 E 7: [command] Command 'remotetemp/11' failed with code: 5
ems-esp:#
`

I've noticed that boiler started to heat 
<img width="1445" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/b1b7709d-a41e-4237-bb7a-9a0f158a664e">

but even after 5 mins no thermostat shown on the EMS ESP UI
<img width="1682" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/5ac81d13-ce90-41c6-9381-1c9ba6e87822">

However I could set FB10 in the FW200 menu for hc1 and it shows the selected temp 11.

Attaching the log just in case.
[log.txt](https://github.com/emsesp/EMS-ESP32/files/14157906/log.txt)

Let me know if you need anything else.

**Reply by vtamac** [reply]
Also even after a couple of minutes FW200 displays a failure as 'type/code 1 remote controller is missing' (translated from Hungarian).
I believe I've read somewhere that I should repeat the temp command every 1 minute or so?

Thanks

**Reply by MichaelDvP** [reply]
The emulation automatic repeats every minute, but the log shows me some  errors. Try the new uploaded version (same version number).
I've fixed the telegram number and ht3 poll.

**Reply by vtamac** [reply]
This is what I get in terminal
`
ems-esp:# call thermostat hc1/remotetemp 20
000+00:07:30.304 I 2: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 20 and id 1 on device 0x10

000+00:07:56.502 W 3: [telegram] Incomplete Rx: 90 88 00 07 00 0B 01 01 00 00 00 00 00 00 00 00 00 00 00 00
ems-esp:# call thermostat hc1/remotetemp 11

000+00:07:56.632 I 4: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 11 and id 1 on device 0x10

000+00:08:15.633 W 5: [telegram] Incomplete Rx: 18 88 88 00 18 00 00 02 2A 64 00 00 02 00 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 00 00
`

log is attached 
[log (2).txt](https://github.com/emsesp/EMS-ESP32/files/14161726/log.2.txt)

**Reply by vtamac** [reply]
Still no new device on ESP UI, but FW200 tells that "there is new remote controller on the network please adjust settings".

**Reply by MichaelDvP** [reply]
Ok, next try, now the reply to requests should work with ht3. Please log a few minutes more.

**Reply by vtamac** [reply]
Terminal
`
000+00:00:26.748 I 0: [shell] User session opened on console pty0
ems-esp:$ su
Password:

000+00:00:36.886 N 1: [shell] Admin session opened on console pty0
ems-esp:# call thermostat hc1/remotetemp 19

000+00:00:45.279 I 2: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 19 and id 1 on device 0x10

000+00:00:55.342 E 3: [telegram] Last Tx Read operation failed after 3 retries. Ignoring request: 8B 90 12 00 1B

000+00:01:12.616 W 4: [telegram] Incomplete Rx: 90 88 00 07 00 0B 01 01 00 00 00 00 00 00 00 00 00 00 00 00

000+00:01:24.930 I 5: [emsesp] Recognized new thermostat with deviceID 0x18

000+00:01:39.885 W 6: [telegram] Incomplete Rx: 98 88 00 18 00 00 01 C1 64 00 00 02 00 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 00 00

000+00:02:10.141 W 7: [telegram] Incomplete Rx: 98 88 88 00 18 00 00 01 C0 64 00 00 02 00 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 00 00

000+00:02:43.664 W 8: [telegram] Incomplete Rx: 88 00 18 00 32 01 BF 64 00 01 03 00 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02

000+00:02:45.540 W 9: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 BF 64 00 01 03 24 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:02:52.495 W 10: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 B2 64 3E 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:03:00.230 W 11: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 67 64 39 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:03:41.051 W 12: [telegram] Incomplete Rx: 88 00 18 00 32 01 2D 64 2B 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02

000+00:04:31.427 W 13: [telegram] Incomplete Rx: 98 88 00 18 00 32 01 2B 64 2B 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18
ems-esp:# call thermostat hc1/remotetemp  30

000+00:05:00.971 I 14: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 30 and id 1 on device 0x10

000+00:05:29.153 W 15: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 45 64 42 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:06:12.160 W 16: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 6F 64 56 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:06:22.172 W 17: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 79 64 58 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:06:32.402 W 18: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 82 64 58 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:07:02.654 W 19: [telegram] Incomplete Rx: 98 88 88 00 18 00 32 01 93 64 5D 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:07:12.059 W 20: [telegram] Incomplete Rx: 98 88 00 07 00 03 01 01 00 00 00 00 00 00 00 00 00 00 00 00

000+00:07:53.524 W 21: [telegram] Incomplete Rx: 98 88 00 18 00 32 01 A7 64 63 09 03 25 C0 80 00 80 00 80 00 FF FF FF 00 00 00 00 00 02 18

000+00:08:10.925 E 22: [telegram] Last Tx Read operation failed after 3 retries. Ignoring request: 8B 90 FF 00 19 00 6F

000+00:08:25.972 W 23: [telegram] Incomplete Rx: 98 88 00 07 00 03 01 01 00 00 00 00 00 00 00 00 00 00 00 00

000+00:08:32.684 W 24: [telegram] Incomplete Rx: 98 88 00 07 00 0B 01 01 00 00 00 00 00 00 00 00 00 00 00 00
`
log is attached. Device now shows up on the UI and picks up the temp sent in [log (3).txt](https://github.com/emsesp/EMS-ESP32/files/14162384/log.3.txt)
<img width="1662" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/b64d4290-3e8b-4480-9526-482035455381">

**Reply by MichaelDvP** [reply]
But it seems the FW200 does not set the roomtemp. Changed to 0x123 again, as seen in #773. 
The incomplete telegrams seems, that sometimes the break is missed after polling the remote, but very random. What TX-mode do you use?

**Reply by MichaelDvP** [reply]
Please update again and test.

**Comment by vtamac** [comment]
<img width="1647" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/135bcdac-f6e8-4ae4-b54f-df0c0102310c">

FB10 shows up again. Log is attached. [log (5).txt](https://github.com/emsesp/EMS-ESP32/files/14166392/log.5.txt)

Also, should I set anything on my FW200 I normally do not change anything on it during tests

**Reply by MichaelDvP** [reply]
This look good: 
`000+00:01:31.876 TRACE 138: [emsesp] thermostat(0x18) -W-> thermostat(0x10), JunkersRemote(0x0123), data: 00 64
000+00:01:32.269 TRACE 139: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01 01 00 32 00 FA 00 64 00`
The remote sends 0064 and the master reports 00FA for internals and  0064 for remote sensor, good.
Now you need to set on the FW200 `control` to FB10 to use the reported roomtemp, it's set to off now (second byte here:
`000+00:01:02.758 TRACE 112: [emsesp] thermostat(0x10) -W-> Me(0x0B), JunkersSet(0x0165), data: 01 00 02 19 4B 32 05 01 00 03 00 33 F6 01 01 0A 28 3C 03 01 00`)

**Reply by MichaelDvP** [reply]
BTW. Afair the FW200 only supports 2 heating circuits with remote FB10, a 3rd hc has to be controlled by a FB100, that acts as complete thermostat (sending 0x171/0x167 telegrams with all parameters). But for now the FB10 logic works for hc1-hc4, so you can try.

**Reply by vtamac** [reply]
Oh okay, well I can live with 2hc's that would be a great upgrade already. What are the steps required for hc2? I've tried sending a command in terminal but gives error.
`
ems-esp:# call thermostat hc2/remotetemp 10
Bad syntax (error code 5)
000+00:23:29.177 I 42: [command] Calling command 'thermostat/remotetemp' (room temperature from remote) with value 10 and id 2 on device 0x00
000+00:23:29.177 E 43: [command] Command 'remotetemp/10' failed with code: 5
`

**Reply by vtamac** [reply]
also after a couple of hours, FW200 started to complain about the missing FB10 again, however, it is still showing up on the webui. It also started to heat even if the last value of the FB10 I set was 40 degrees. As it feels that FB10 is missing it is using its on temp maybe?
What is also strange the even despite fw200 showing an error it picks up the tmp from FB10 but looks like not considering when setting starting to heat
<img width="1426" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/2cdeb69b-8722-412b-baa2-53311feae5b1">
<img width="1423" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/0f90e2db-4232-46c8-b000-12d45e060dd0">

**Reply by MichaelDvP** [reply]
- Please tell the error message and attach a log.
- For hc 2 you first have to configure the FW200 to control the 2.nd hc, Do you have a mixer/pump-unit for the 2nd hc? Havn't seen it in devices. After that you can send the remotetemp command for hc2.
- Heating is a slow process, the thermostat calcualte the right flowtemp and tells it to the boiler. The boiler tries to make this flowtemp, starts heating if the actual flowtemp is lower (with hysteresis) and stops (with hysteresis) if flowtemp is above setpoint. Here 50°C should be reached and actal is 41,5°C, so the boiler heats up. (Stops heating at ~60° and restarts at ~40 current flowtemp).

**Reply by vtamac** [reply]
1.  Will do a clean setup from the ground and will let you know about the results
2. As far the heating I'm only suprised because the selected room temp is 30 but the current temp set to 39 and fw200 still commands the boiler to heat

**Reply by vtamac** [reply]
HI @MichaelDvP 

So I have created a clean setup, enabled HC1 and HC2 on FW200 and called call thermostat hc1/remotetemp xx and call thermostat hc2/remotetemp xx in shell. Both new FB100 is displayed on the web UI, however, FW200 displays an error. "missing bus component IPM". I've set both HC in FW200 menu to mixed. The available hc types are the following 

1. Not-mixed with IPM
2. Not-mixed without IPM
3. Mixed

Unfortunately I cannot set Not-Mixed without IPM for both HCs as FW200 displays an error that two not-mixed hc without ipm is not supported. 
Please find a log attached.
[log (6).txt](https://github.com/emsesp/EMS-ESP32/files/14233276/log.6.txt)

Thanks,
Tamás

**Reply by MichaelDvP** [reply]
I thought you have already 2 heatingcircuits. 
Sure the heating needs hardware for each circuit: a pump and a control for the pump, this is the IPM module. Without that you cant realize a heatingcircuit.

**Reply by vtamac** [reply]
Sorry for the misunderstanding! Unfortunately, You can't buy these modules anymore (IPM, FB10). I might be able to solve my issues with 1 heating circuit - at least there will be one which FW200 knows about and I can control the emulated FB100 via home assistant and update it with the values of the circuit which requires more heat (where the desired and actual temp diff is the biggest). 
If going with 1 circuit and one FB10 what are the next steps required?

1. Is your branch going to be merged into the main so that I can upload the new releases too (if any)?
2. What's the process of this FB10 emulation to remain permanent? If the ESMESP board reboots I have to call this function again?


Thank you,
Tamás

**Reply by vtamac** [reply]
Actually, even with 1 HC the FW200 shows an error about a missing bus component (FB10).
<img width="1650" alt="image" src="https://github.com/emsesp/EMS-ESP32/assets/17232589/ff33e59e-38a9-42d6-8834-7b07c07ef0f8">

Please find the log attach
[log (7).txt](https://github.com/emsesp/EMS-ESP32/files/14258543/log.7.txt)
ed.

**Reply by MichaelDvP** [reply]
Seems to be the same as in #1611, maybe sending more often helps, i've changed in my test build.

I think IPM are sometimes on ebay, but you also need a pump and change the piping to split the heatingcircuit. That's a lot of work to do.

**Reply by proddy** [reply]
See wiki article on this: https://emsesp.github.io/docs/Special-Functions/#remote-thermostats
